LCC Credit: Add comchain credit, improve cyclos credit, small lint and fixes

njeudy opened the pull request:
Implement entrypoint for comchain:

- credit
- credit-requests
- validate-credit-request

Commit:
[FIX] lcc_cyclos_base: improve credit

- Transfered to principal module
- Improve paid invoice selector for cyclos

vaab commented:
For the record:

Create first commit with only black:
```
git co -b black origin/12.0-Lokavaluto-app-connection
black .
git add -A lcc*
git ci -m "[CHG] dev: lint code with black"
```
Then:
```
git co -b rebuild_pr 1b8792d      ## first commit of the pr
git reset black
```
And `git diff` shows a lot of inconsistencies: code added and file not linted with black.

vaab committed:
[NEW] dev: add ``comchainCreateOrder`` method in ``res.partner``

- implement comchain credits api entrypoints:
- credit
- credits-requests
- validate-credit-request

vaab commented on lcc_lokavaluto_app_connection/services/partner_services.py:
```diff
@@ -51,6 +51,23 @@ def backend_credentials(self):
+        [(["/validate-credit-request"], "POST")],
+        input_param=Datamodel("partner.validate.credit.requests.param"),
+    )
+    def credit_requests(self, partner_credit_requests_get_param):
```
second function having the same name.

njeudy replied:
ok I see, will change this :)

vaab commented on lcc_comchain_base/services/comchain_services.py:
```diff
@@ -74,3 +74,24 @@ def activate(self, params):
+        output_param=Datamodel("comchain.credit.response"),
+    )
+    def credit(self, params):
+        """
+        Credit user account with amount, and generate accounting entry
+        """
```
minor: not PEP257 https://peps.python.org/pep-0257/ it is best on one line ( https://peps.python.org/pep-0257/#one-line-docstrings ). And BTW, you can remove the comma (",").

njeudy replied:
Ok done.

vaab commented on lcc_comchain_base/views/account_invoice_view.xml:
```diff
@@ -0,0 +1,16 @@
+        <field name="arch" type="xml">
+            <xpath expr="//sheet" position="before">
+                <div groups="account.group_account_invoice" class="alert alert-info" role="alert" style="margin-bottom:0px;" attrs="{'invisible': [('comchain_amount_to_credit','=',0)]}">
+                    You have <bold><field name="comchain_amount_to_credit"/></bold> numeric currency to credit for this customer. Please go to mobile app to validate credit.
```
"Please use" instead of "Please go to" seems more natural. And the last "credit" would maybe clearer if called "credit requests".

njeudy replied:
Ok done

vaab commented on lcc_lokavaluto_app_connection/datamodel/partner_info.py:
```diff
@@ -11,6 +11,12 @@ class PartnerInfoGetParam(Datamodel):
     backend_keys = fields.List(fields.String())
 
 
+class PartnerCreditRequestsGetParam(Datamodel):
```
Didn't find a place were you use this.

vaab commented on lcc_lokavaluto_app_connection/models/account_invoice.py:
```diff
@@ -0,0 +1,32 @@
+    )
+
+    @api.one
+    @api.depends("state", "invoice_line_ids.product_id", "invoice_line_ids.price_total")
```
not sure what depends do... so not sure why depend on `price_total` ?

njeudy replied:
Hum copy/paste I think

vaab commented on lcc_lokavaluto_app_connection/models/account_invoice.py:
```diff
@@ -0,0 +1,32 @@
+        if lcc_numeric_products:
+            self.has_numeric_lcc_products = True
+
+    def _get_credit_requests(self, backend_keys):
```
a small comment to help recall that these function are to be implemented by backends modules may be helpful.

njeudy replied:
Ok done :)

vaab commented on lcc_comchain_base/datamodel/comchain.py:
```diff
@@ -0,0 +1,71 @@
+    type = fields.Integer(required=True)
+    credit_min = fields.Float(required=True)
+    credit_max = fields.Float(required=True)
+
+
+class ComchainActivateList(Datamodel):
+    _name = "comchain.activate.list"
+
+    accounts = fields.List(NestedModel("comchain.activate.info"), required=True)
+
+
+class ComchainAccountInfo(Datamodel):
+    _name = "comchain.account"
+
+    partner_id = fields.Integer(required=True)
+    name = fields.String(required=True)
+    comchain_address = fields.String(required=True)
+
+
+class ComchainCreditInfo(Datamodel):
+    _name = "comchain.credit.info"
+
+    comchain_address = fields.String(required=True)
+    amount = fields.Float(required=True)
+
+
+class ComchainCreditResponse(Datamodel):
+    _name = "comchain.credit.response"
+
+    order_url = fields.String(required=True)
+
+
+class ComchainCreditRequest(Datamodel):
+    _name = "comchain.credit.request"
+
+    credit_id = fields.Integer(required=True)
+    amount = fields.Float(required=True)
+    date = fields.Date(required=True)
+    name = fields.String(required=True)
+    monujo_backend = fields.List(fields.String(), required=True)
```
It seems this file is based on `partners_info.py` (nearly identical). I guess there might be an issue here with wanting to rename a file or something like that. I might not have understood completely the need for this file or the existence of `partners_info.py` in the same directory.

vaab commented on lcc_comchain_base/datamodel/comchain.py:
```diff
@@ -0,0 +1,71 @@
+from odoo.addons.datamodel.fields import NestedModel
+
+
+class AddressesInfo(Datamodel):
```
can't find a place where this would be used. Do you want to keep it for coherence between modules ?

Commit:
[IMP] lcc_lokavaluto_app_connection small improvment

vaab commented:
cleaned history and reviewed by phone, pushed in 12.0-Lokavaluto-app-connection
